src/ts_catalog/chunk_rewrite.c

```diff
@@ -0,0 +1,297 @@
+			return ChunkRewriteOngoing;
+	}
+
+	/*
+	 * Check if the new heap still exists by trying to get a lock.
+	 */
+	Relation newrel = try_table_open(form.new_relid, AccessExclusiveLock);
+
+	if (newrel)
+	{
+		ObjectAddress tableaddr;
+		/* New heap still exists, so delete it */
+		table_close(newrel, NoLock);
+		ObjectAddressSet(tableaddr, RelationRelationId, form.new_relid);
+		performDeletion(&tableaddr, DROP_RESTRICT, PERFORM_DELETION_INTERNAL);
+		result = ChunkRewriteEntryDeletedAndTableDropped;
+	}
+	else
+	{
+		result = ChunkRewriteEntryDeleted;
+	}
+
+	ts_chunk_rewrite_delete_by_tid(&tid);
+
+	return result;
+}
+
+TS_FUNCTION_INFO_V1(ts_chunk_rewrite_cleanup);
+
+/*
+ * Clean up failed chunk rewrites.
+ *
+ * This function cleans up all non-ongoing chunk rewrites listed in the chunk_rewrite catalog,
+ * including any "orphaned" heaps referenced in the table.
+ *
+ */
+Datum
+ts_chunk_rewrite_cleanup(PG_FUNCTION_ARGS)
+{
+	ScanIterator it;
+	ObjectAddresses *objaddrs = new_object_addresses();
+	unsigned int cleanup_count = 0;
+	unsigned int skipped_count = 0;
+	CatalogSecurityContext sec_ctx;
+	ScanTupLock tuplock = {
+		.lockmode = LockTupleExclusive,
+		.waitpolicy = LockWaitSkip,
+		.lockflags = TUPLE_LOCK_FLAG_FIND_LAST_VERSION,
+	};
+
+	it = ts_scan_iterator_create(CHUNK_REWRITE, RowShareLock, CurrentMemoryContext);
+	it.ctx.tuplock = &tuplock;
+
+	ts_scanner_foreach(&it)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&it);
+		bool isnull = false;
+		bool entry_cleaned = false;
+
+		Datum chunk_relid_dat = slot_getattr(ti->slot, Anum_chunk_rewrite_chunk_relid, &isnull);
+		Assert(!isnull);
+		Datum new_relid_dat = slot_getattr(ti->slot, Anum_chunk_rewrite_new_relid, &isnull);
+		Assert(!isnull);
+
+		if (ti->lockresult == TM_Ok)
+		{
+			Oid new_relid = DatumGetObjectId(new_relid_dat);
+
+			/*
+			 * A concurrent merge might be in progress, so try to lock the "new"
+			 * relation and only delete it if the lock can be acquired
+			 * immediately. If a lock cannot be acquired, a merge is probably
+			 * ongoing and it might still complete successfully.
+			 */
+			if (ConditionalLockRelationOid(new_relid, AccessExclusiveLock))
+			{
+				ObjectAddress new_objaddr = {
+					.objectId = DatumGetObjectId(new_relid),
+					.classId = RelationRelationId,
+				};
+
+				add_exact_object_address(&new_objaddr, objaddrs);
+
+				ItemPointer tid = ts_scanner_get_tuple_tid(ti);
+
+				ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+				ts_catalog_delete_tid_only(it.ctx.tablerel, tid);
+				ts_catalog_restore_user(&sec_ctx);
+				entry_cleaned = true;
+			}
+		}
+
+		if (entry_cleaned)
+		{
+			cleanup_count++;
+		}
+		else
+		{
+			Oid chunk_relid = DatumGetObjectId(chunk_relid_dat);
+			elog(DEBUG1, "chunk merge in progress for \"%s\", skipping", get_rel_name(chunk_relid));
+			skipped_count++;
+		}
+	}
+
+	ts_scan_iterator_close(&it);
+	performMultipleDeletions(objaddrs, DROP_RESTRICT, 0);
```

Comment:
Can `PERFORM_DELETION_INTERNAL` flag be used as in `merge_chunks_finish()`?

Reply:
Good question. I think probably none of these cases should fire event triggers (which I think is what the flag is disabling), since we aren't in fact dropping any data. Still, we do drop a chunk that is no longer needed because its data was merged into another chunk.

But since data stays the same, perhaps we should never fire even triggers? I am in favor of adding the internal flag to all deletes... WDYT?

Reply:
Made all drops use PERFORM_DELETION_INTERNAL for this case.